invokeai/app/services/session_processor/session_processor_default.py

```diff
@@ -7,135 +8,262 @@
+                        graph_execution_state_id=graph_execution_state_id, output_path=stats_path
+                    )
+                # We'll get a GESStatsNotFoundError if we try to log stats for an untracked graph, but in the processor
+                # we don't care about that - suppress the error.
+                with suppress(GESStatsNotFoundError):
+                    self._invoker.services.performance_statistics.log_stats(graph_execution_state_id)
+                    self._invoker.services.performance_statistics.reset_stats()
+
             while not stop_event.is_set():
                 poll_now_event.clear()
+                # Middle processor try block; any unhandled exception is a non-fatal processor error
                 try:
-                    # do not dequeue if there is already a session running
-                    if self.__queue_item is None and resume_event.is_set():
-                        queue_item = self.__invoker.services.session_queue.dequeue()
-
-                        if queue_item is not None:
-                            self.__invoker.services.logger.debug(f"Executing queue item {queue_item.item_id}")
-                            self.__queue_item = queue_item
-                            self.__invoker.services.graph_execution_manager.set(queue_item.session)
-                            self.__invoker.invoke(
-                                session_queue_batch_id=queue_item.batch_id,
-                                session_queue_id=queue_item.queue_id,
-                                session_queue_item_id=queue_item.item_id,
-                                graph_execution_state=queue_item.session,
-                                workflow=queue_item.workflow,
-                                invoke_all=True,
+                    # Get the next session to process
+                    self._queue_item = self._invoker.services.session_queue.dequeue()
+                    if self._queue_item is not None and resume_event.is_set():
+                        self._invoker.services.logger.debug(f"Executing queue item {self._queue_item.item_id}")
+                        cancel_event.clear()
+
+                        # If profiling is enabled, start the profiler
+                        if profiler is not None:
+                            profiler.start(profile_id=self._queue_item.session_id)
+
+                        # Prepare invocations and take the first
+                        invocation = self._queue_item.session.next()
+
+                        # Loop over invocations until the session is complete or canceled
+                        while invocation is not None and not cancel_event.is_set():
+                            # get the source node id to provide to clients (the prepared node id is not as useful)
+                            source_invocation_id = self._queue_item.session.prepared_source_mapping[invocation.id]
```

Comment:
Could make sense to put the code in this loop that actually runs an invocation into a util function or function on the base class that all other session processors can use. Maybe even the whole loop with callback functions as parameters before/after each loop

Reply:
I plan to separate the loop in a followup PR, this one is already getting chunky.

We have a minimal events implementation via `fastapi-events`, but this library supports some pretty nifty stuff, like dispatchable pydantic models for events. Adding structure to the events is a big DX win:

```py
if isinstance(InvocationStartedEvent, event):
    # we now have type hints for `event`
```

I've been planning to refactor the events like this anyways, just haven't gotten around to it.

Do you have any opposition to leaning into events? Callbacks are probably easier to understand and debug but I feel like we may as well use events.